internal/server/least_connections.go

```diff
@@ -0,0 +1,25 @@
+package server
+
+import (
+	"errors"
+
+	"tcp-load-balancer/internal/upstream"
+)
+
+// LeastConnections returns an authorized host with the fewest open connections.
+func (l *LoadBalancer) LeastConnections() (*upstream.TcpHost, error) {
+	if l == nil || len(l.hosts) == 0 {
+		return nil, errors.New("no upstream hosts available")
+	}
+
+	var host *upstream.TcpHost
+
+	for _, h := range l.hosts {
+		// TODO: handle authorization scheme here in next PR. For now, we'll just assume the client can access all hosts.
+		if host == nil || h.ConnectionCount() < host.ConnectionCount() {
```

Comment:
This one is too easy to confuse variable names. I'll rename this to make it clearer.

In this case, `h` comes from the hosts lists, while `host` was initialized as nil on line 15: `var host *upstream.TcpHost` so it will have a `ConnectionCount` of 0, and always be nil until after the first iteration through this loop. 

I'll rename `host` to `selectedHost` so that the intent is clearer.